File: R_src/LASSO.R
LASSO = function(fulldata, export = F, fit_svm = F, fit_rf = F, foldfreq = 0.6) {
  retenv = new.env()
  #############################################################

  lasso_models = list()
  
  auc_RF = c()
  auc_RF_permute = c()
  auc_svm = c()
  auc_svm_permute = c()
  
  
  #This part is process for the lasso regression
  auc_data = c()
  auc_data_permute = c()
  
  #Create the variable to hold values:
  
  # this is the default for createFolds anyway
  k = 10
  numRuns = 10
  
  for(runCount in 1:numRuns) {
    
    selFeat1 <- c()
    compSelFeat1 <- c()
    chosenFeat1 <- c()
    total_oriY <- c()
    total_oriYhat <- c()
    
    selFeat2 <- c()
    compSelFeat2 <- c()
    chosenFeat2 <- c()
    total_oriY <- c()
    total_permYhat <- c()
    
    append.Y = c()
    append.Y_permute = c()
    
    append_RF = c()
    append_RF_permute = c()
    append_svm = c()
    append_svm_permute = c()
    
    trueY = c()
    permuteY = c()
    selectedVars = c()
    total_permY = c()
    
    # scramble response vector
    permutedY = sample(fulldata$Group)
    
    #create kfold associate to a data frame.
    kfolds <- createFolds(y = fulldata$Group, k=k, list = FALSE, returnTrain = FALSE)
    myData <- cbind(fulldata, kfolds)
    
    ##############################################
    
    for (select in 1:k) {
      #Create different subdata by different fold.
      chooseFold <- which(kfolds == select)
      
      # copy original data
      newData = myData
      
      for (n_run in 1:2) {
        
        # select training data
        
        #take out the random row in newData and put into train, 
        #and the test data only contain the one which the train data does not have
        train <- newData[-chooseFold, ]
        test <- newData[chooseFold, ]
        
        #remove the kfold column which is the last column out of the data to calculate
        train <- train[, -ncol(train)]
        test <- test[, -ncol(test)]
        
        #Since X data will be constant
        X_train <- as.matrix(train[, -1])
        
        # not used until later
        X_test <- as.matrix(test[, -1])
        
        y = train$Group
        
        if (n_run == 2) {
          newData$Y = permutedY
        }
        
        all_signif_vars = c()
        
        for (num_FS in 1:numRuns) {
          
          # run cv lasso to find best lambda
          lasso_run <- cv.glmnet(x=X_train, y=y, alpha=1, nfolds = k,
                                 standardize = TRUE)
          
          #run lasso on training data with best lambda
          lasso_run_lmin <- glmnet(x=X_train, y=y, alpha=1, 
                                   lambda = lasso_run$lambda.min)
          
          c = coef(lasso_run_lmin)
          inds = which(c!=0)
          
          #remove intercept from the list
          tmp_variables <- tail(row.names(c)[inds], -1)
          #print(tmp_variables)
          
          all_signif_vars<-c(all_signif_vars,tmp_variables)
          # save features
          
          # lasso_predict1 = predict(lasso_run_lmin, newx = X_test)
        }
        
        # select features based on frequency
        freq = sort(table(all_signif_vars),decreasing=TRUE)/numRuns
        
        # only pick features which show up above frequency threshold
        variables = names(which(freq > foldfreq))
        
        # variables = all_signif_vars
        #subset train and test data correspond to the chosen features
        selTrain <- train[, (names(train) %in% variables)]
        newTrain <- cbind.data.frame(train$Group, selTrain)
        
        selTest <- test[, (names(test) %in% variables)]
        newTest <- cbind.data.frame(test$Group, selTest)
        
        colnames(newTrain)[1] <- "Y"
        colnames(newTest)[1] <- "Y"
        
        # for true Y and permuteY, put in separate lists but both named Y
        if (n_run == 1) {
          trueY = append(trueY, newTest$Y)
          Y = trueY
        } else {
          permuteY = append(permutedY, newTest$Y)
          Y = permutedY
        }
        
        if (fit_svm){
        # svm
          svmfit = svm(y= newTrain$Y, x=as.matrix(newTrain[,-1]), kernel="linear", cost=10, scale=FALSE)
          yhat.SVM = predict(svmfit, newdata = newTest[, -1], type = "response")
        }
        
        if (fit_rf) {
        # random forest
          RFfit <- randomForest(y= newTrain$Y, x=as.matrix(newTrain[,-1]), importance=TRUE, ntree = 100)
          yhat.RF = predict(RFfit, newdata = newTest[, -1], type = "response")
        }
        
        # measuring yhat
        lasso_run2 <- glmnet(y= newTrain$Y, x=as.matrix(newTrain[,-1]), alpha=1,
                             lambda = lasso_run$lambda.min)
        
        yhat <-predict(lasso_run2, newx=as.matrix(newTest[,-1], type = "response"))
        
        # store yhat in appropriate location depending on mode
        if (n_run == 1) {
          append.Y = append(append.Y, yhat)
          selFeat1 <- c(selFeat1, variables)
          # compSelFeat1 <- c(compSelFeat1,selFeat1[select == k])
          compSelFeat1 <- c(compSelFeat1,selFeat1)
          
          
          lasso_models[[runCount]] = list(fit = lasso_run2, variables = variables)
          #svm and rf
          if (fit_svm) {
            append_svm = c(append_svm, yhat.SVM)
            
          }
          if (fit_rf) {
            append_RF = c(append_RF, yhat.RF)
          }
          
        } else {
          append.Y_permute = append(append.Y_permute, yhat)
          selFeat2 <- c(selFeat2, variables)
          # compSelFeat2 <- c(compSelFeat2,selFeat2[select == k])
          compSelFeat2 <- c(compSelFeat2,selFeat2)
          
          #svm and rf
          if (fit_svm) {
            append_svm_permute = c(append_svm_permute, yhat.SVM)
            
          }
          if (fit_rf) {
            append_RF_permute = c(append_RF_permute, yhat.RF)
          }
        }
      }
      chosenFeat1 <- c(chosenFeat1,compSelFeat1)
      chosenFeat2 <- c(chosenFeat2,compSelFeat2)
      
    }
    if(length(unique(permutedY))>2){
      #measure the classification accuracy
      total_oriY <- c(total_oriY, append.Y)
      total_oriYhat <- c( total_oriYhat, trueY)
      
      total_permY <- c(total_permY, append.Y_permute)
      total_permYhat <- c(total_permYhat, permutedY)
    }
    
    # save AUC scores
    auc_data[runCount] = auc(trueY, append.Y)
    auc_data_permute[runCount] = auc(permutedY, append.Y_permute)
    
    if (fit_rf) {
      auc_RF[runCount] = auc(trueY, append_RF)
      auc_RF_permute[runCount] = auc(permutedY, append_RF_permute)
    }
    
    if (fit_svm) {
      auc_svm[runCount] = auc(trueY, append_svm)
      auc_svm_permute[runCount] = auc(permutedY, append_svm_permute)
    }
  }
  
  classifiedAccuracy <- confusionMatrix(as.factor(trueY), 
                                        as.factor(round(append.Y)))
  
  classifiedAccuracyPerm <- confusionMatrix(as.factor(permutedY), 
                                            as.factor(round(append.Y_permute)))
  
  # to get freq features as needed
  cf_ori = sort(table(compSelFeat1), decreasing = T)
  # cf_ori = names(which(cf_ori > foldfreq * ))
  
  chosenFeat_ori <- table(chosenFeat1) %>% as.data.frame() %>% arrange(desc(Freq))
  chosenFeat_perm <- table(chosenFeat2) %>% as.data.frame() %>% arrange(desc(Freq))
    if (export) {
      
    write.table(chosenFeat_ori, 
                file="output/chosenFeat_ori", 
                append = FALSE, sep = " ",
                row.names = FALSE, col.names = TRUE) 
    write.table(chosenFeat_perm, 
                file="output/chosenFeat_perm2", 
                append = FALSE, sep = " ",
                row.names = FALSE, col.names = TRUE) 
    
    # make dataframes for classification info
    class_accur = data.frame(cbind(t(classifiedAccuracy$byClass)))
    
    class_accurperm = data.frame(cbind(t(classifiedAccuracyPerm$byClass)))
    
    write.table(class_accur, 
                file="output/classifiedAccuracy", 
                append = FALSE, sep = " ",
                row.names = FALSE, col.names = TRUE) 
    write.table(class_accurperm, 
                file="output/classifiedAccuracyPerm", 
                append = FALSE, sep = " ",
                row.names = FALSE, col.names = TRUE) 
    
  }
  
  retenv$feats = list(chosenFeat = chosenFeat_ori, compSelFeat = cf_ori, neg = chosenFeat_perm)
  retenv$auc = list(lasso_auc = list(pred = auc_data, perm = auc_data_permute), 
                    svm_auc = list(pred = auc_svm, perm = auc_svm_permute), 
                    rf_auc = list(pred = auc_RF, perm = auc_RF_permute))
  
  retenv$models = lasso_models
  return(retenv)
}

File: src/LASSO.py
import pandas as pd
import numpy as np
from sklearn import metrics
from sklearn.metrics import roc_curve, auc
import math
from sklearn.metrics import confusion_matrix
import types
import random
from sklearn.model_selection import KFold
from sklearn.linear_model import Lasso
from sklearn import svm

def calculateAUC(trueY, permutedY,append_model, append_model_permute, unique_levels = 2):

   if len(pd.unique(trueY)) == 2:
    
    auc_model = auc(trueY, append_model)
  
    # auc_model_permute = auc(permutedY, append_model_permute)
    auc_model_permute = metrics.auc(permutedY, append_model_permute)
  
   else:
      auc_model = auc(roc_curve(trueY, append_model))
    
    # auc_model_permute = auc(multiclass.roc(permutedY, append_model_permute))
      auc_model_permute = auc(roc_curve(permutedY, append_model_permute))

   def normbetween(x, unique_levels) :
    # the number of levels is the range we want
    
      if (unique_levels <= 2) :
        return( 1 / (1 + math.exp(-x)))
    
      else :
        norm_factor = 1/(unique_levels - 1)
      
        return( (1 / (norm_factor + math.exp(-x))) + 1 )

   cfm = confusion_matrix(pd.Categorical(round(append_model)),
                       pd.Categorical(trueY))

   cfm_permute = confusion_matrix(pd.Categorical(round(append_model_permute)),
                               pd.Categorical(permutedY))

   return(list(auc_model = auc_model, auc_model_permute = auc_model_permute,
              cfm = cfm, cfm_permute = cfm_permute))

def LASSO_Grid(fulldata, export = F, foldfreq = 0.6, alphaValues = [1.0]) :
  retenv = types.SimpleNamespace()
  #############################################################

  random.seed(100)
  alphaValues = float(alphaValues)
  
  # number of folds to create
  k = 10
  
  # number of repeats 
  numRuns = 10
  
  # initialize results list
  gridResults = list(numRuns)
  ##################################################################################################################################
  for runCount in 1:numRuns:
    
    gridValues = list()
    for val in alphaValues:
      gridValues[[len(gridValues) + 1]] = list(alpha = val)
    
    
#create kfold associate to a data frame.
  z = fulldata['Group']

# specify the number of folds
  k = 5

# create KFold object
  kf = KFold(n_splits=k, shuffle=True, random_state=42)

# create an empty list to store the folds
  folds = []

# loop through the folds
  for train_index, test_index in kf.split(z):
    # append the test indices to the folds list
    folds.append(test_index)

# print the folds
  print(folds)

  mydata = pd.concat([fulldata, folds], axis=1)    
    # scramble response vector
  permutedY = random.sample(z)
    
  unique_levels = len(pd.unique(z))
    
    ##############################################
    
  for fold in range(1,k) :
      #Create different subdata by different fold.
      currentFold = np.where(folds == fold)
      
      # copy original data
      newData = myData
      variables = list()
  ################################################################################################################################### 
      for entry in range(1,len(gridValues)) :

        for n_run in range(1,2) :
          #take out the random row in newData and put into train, 
          #and the test data only contain the one which the train data does not have
          ########################################################################################################################
          train = newData[-currentFold, ]
          test = newData[currentFold, ]
          ########################################################################################################################
          #remove the kfold column which is the last column out of the data to calculate
          train = train[:, :-1]
          test = test[:, :-1]
          ########################################################################################################################
          #Since X data will be constant
          X_train = train.iloc[:, 1:].values

          # not used until later
          X_test = test.iloc[:, 1:].values
          
          y = train['Group']
          ######################################################################################################################
          # run cv lasso to find best lambda
        
          lasso_run = LassoCV(cv=k, normalize=True, fit_intercept=False)
          lasso_run.fit(X_train, y)

          alpha_value = lasso_run.alpha_

          lasso_run_lmin = Lasso(alpha=alpha_value*gridValues[entry]['alpha'], fit_intercept=False, normalize=False)
          lasso_run_lmin.fit(X_train, y)

          c = lasso_run_lmin.coef_
          inds = np.where(c != 0)[0]
          #######################################################################################################################
           

            #remove intercept from the list
          variables[entry] = list(c.index[inds][1:])

          selTrain = train.loc[:, train.columns.isin(variables[entry])]
          newTrain = pd.concat([train['Group'], selTrain], axis=1)
            
          selTest = test.loc[:, test.columns.isin(variables[entry])]
          newTest = pd.concat([test['Group'], selTest], axis=1)
            
          newTrain = newTrain.rename(columns={newTrain.columns[0]: "Y"})
          newTest = newTest.rename(columns={newTest.columns[0]: "Y"})
          
          # for true Y and permuteY, put in separate lists but both named Y
          if (n_run == 1) :
              gridValues[entry]['trueY'] = gridValues[entry]['trueY'] + newTest['Y'].tolist()
              
          
          else : 
              gridValues[entry]['permuteY'].append(newTest['Y'])
              
            
              newTrain = newTrain.droplevels()
              newTest = newTest.droplevels()

          svmfit = svm.SVC(kernel='linear', C=10, probability=True)
          svmfit.fit(np.array(newTrain.drop('Y', axis=1)), np.array(newTrain['Y']))

          # make predictions on new data
          yhat_SVM = svmfit.predict_proba(np.array(newTest.drop('Y', axis=1)))[:, 1]
          # random forest
          RFfit = RandomForestClassifier(n_estimators=100, random_state=0)
          RFfit.fit(X=newTrain.iloc[:, 1:], y=newTrain['Y'])

          # Making predictions on the test set
          yhat_RF = RFfit.predict(newTest.iloc[:, 1:])

            # store yhat in appropriate location depending on mode
          if (n_run == 1) {
              gridValues[entry]['chosenFeats'] = variables[entry]
              
              #svm and rf
              gridValues[entry]['append_svm'].extend(yhat_SVM)
              gridValues[entry]['svm_models'] = {'fit': svmfit, 'variables': variables[entry]}

              # Appending the predictions of the random forest and the fitted model to the corresponding lists in the dictionary
              gridValues[entry]['append_RF'].extend(yhat_RF)
              gridValues[entry]['rf_models'] = {'fit': RFfit, 'variables': variables[entry]}

          else :
             # Storing the selected variables in the dictionary
             gridValues[entry]['chosenFeat_perm'] = variables[entry]

             # Appending the predictions of the SVM and the random forest to the corresponding lists in the dictionary
             gridValues[entry]['append_svm_permute'].extend(yhat_SVM)
             gridValues[entry]['append_RF_permute'].extend(yhat_RF)  

          for (entry in range(1,len(gridValues)):
      
# Calculating the AUC for the SVM and the random forest
              gridValues[entry]['svm'] = calculateAUC(gridValues[entry]['trueY'],
                                         permutedY,
                                         gridValues[entry]['append_svm'],
                                         gridValues[entry]['append_svm_permute'],
                                         unique_levels)

              gridValues[entry]['rf'] = calculateAUC(gridValues[entry]['trueY'],
                                        permutedY,
                                        gridValues[entry]['append_RF'],
                                        gridValues[entry]['append_RF_permute'],
                                        unique_levels)

    

    gridResults[[runCount]] = gridValues
  

  retenv['gridResults'] = gridResults

  return(retenv)
   
